### PR TITLE
[Core] Nodal position expression IO

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR adds `read` and `write` between `ModelPart::Node` initial and current configurations (nodal positions). A proper test also added and tested with 140 mpi cores. 

**🆕 Changelog**
- Added `NodalPositionExpressionIO`
- Added tests

### CODE REVIEW

**Review by RiccardoRossi — COMMENTED** [review]
Hi Suneth, please put some documentationof this somewhere...otherwise we will never learn how to use this

File: kratos/expression/expression_io_utils.h [hunk 1]
```diff
@@ -219,6 +222,87 @@ class ExpressionIOUtils
         }
     }
 
+    /**
+     * @brief Evaluates a local expression on ghost nodes.
+     *
+     * This method allows evaluating a local expression (@ref rLocalNodesExpression) (expressions are by default only works on local mesh) on
```
Comment (matekelemen) [inline kratos/expression/expression_io_utils.h]:
```suggestion
     * This method allows evaluating a local expression (@a rLocalNodesExpression) (by default, expressions act only on the local mesh) on
```

File: kratos/expression/expression_io_utils.h [hunk 2]
```diff
@@ -219,6 +222,87 @@ class ExpressionIOUtils
...
+     * the ghost nodes. @ref TGhostNodesContainerType can be either const ModelPart::NodesContainerType or ModelPart::NodesContainerType. Depending on
```
Comment (matekelemen) [inline kratos/expression/expression_io_utils.h]:
```suggestion
     * the ghost nodes. @a TGhostNodesContainerType can be either const @a ModelPart::NodesContainerType or @a ModelPart::NodesContainerType. Depending on
```

File: kratos/expression/expression_io_utils.h [hunk 3]
```diff
@@ -219,6 +222,87 @@ class ExpressionIOUtils
...
+     * the constness of the @ref TGhostNodesContainerType, the signature of the @ref rApplyFunctor differs.
+     *      If @ref TGhostNodesContainerType is const ModelPart::ContainerType -> void(const Node& rGhostNode, const std::vector<double>& rGhostNodeValues)
+     *      If @ref TGhostNodesContainerType is ModelPart::ContainerType -> void(Node& rGhostNode, const std::vector<double>& rGhostNodeValues)
```
Comment (matekelemen) [inline kratos/expression/expression_io_utils.h]:
```suggestion
     *      If @a TGhostNodesContainerType is const ModelPart::ContainerType, @a TApplyFunctor must be void(const Node& rGhostNode, const std::vector<double>& rGhostNodeValues)
     *      If @a TGhostNodesContainerType is ModelPart::ContainerType. @a TApplyFunctor must be void(Node& rGhostNode, const std::vector<double>& rGhostNodeValues)
```

File: kratos/expression/expression_io_utils.h [hunk 4]
```diff
@@ -219,6 +222,87 @@ class ExpressionIOUtils
...
+     * @param rGhostNodes                       Ghost nodes container.
+     * @param rApplyFunctor                     Apply funktor to do something with each expression values evaluated for ghost nodes.
+     */
+    template<class TGhostNodesContainerType, class TApplyFunctor>
+    static void EvaluateExpressionOnGhostNodes(
+        const DataCommunicator& rDataCommunicator,
+        const Expression& rLocalNodesExpression,
+        const ModelPart::NodesContainerType& rLocalNodes,
+        TGhostNodesContainerType& rGhostNodes,
+        TApplyFunctor&& rApplyFunctor)
```
Comment (matekelemen) [inline kratos/expression/expression_io_utils.h]: This signature is wide open to misuse. A user could pass all these objects from different `ModelPart`s, which would definitely lead to all sorts of exotic bugs.

How about passing a `Communicator`, an `Expression`, and a `TApplyFunctor`? The rest can be fetched from the `Communicator`.

File: kratos/expression/nodal_position_expression_io.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    switch (mConfigurationType) {
+        case ConfigurationType::Initial:
+            expression = LiteralFlatExpression<double>::Create(number_of_nodes, {3});
+            IndexPartition<IndexType>(number_of_nodes).for_each([&r_mesh, &expression](const IndexType Index) {
+                const auto& r_node = *(r_mesh.NodesBegin() + Index);
+                double* current_data_start = expression->begin() + Index * 3;
+                *(current_data_start++) = r_node.X0();
+                *(current_data_start++) = r_node.Y0();
+                *(current_data_start) = r_node.Z0();
```
Comment (matekelemen) [inline kratos/expression/nodal_position_expression_io.cpp]:
```suggestion
                *current_data_start++ = r_node.X0();
                *current_data_start++ = r_node.Y0();
                *current_data_start = r_node.Z0();
```

File: kratos/python/add_expression_io_to_python.cpp [hunk 6]
```diff
@@ -362,6 +362,22 @@ void AddExpressionIOToPython(pybind11::module& rModule)
              pybind11::arg("variable"),
              pybind11::arg("container_type"))
         ;
+
+    auto nodal_expression_io = rModule.def_submodule("NodalPositionExpressionIO");
+    pybind11::enum_<NodalPositionExpressionIO::ConfigurationType>(nodal_expression_io, "ConfigurationType")
+        .value("Initial", NodalPositionExpressionIO::ConfigurationType::Initial)
+        .value("Current", NodalPositionExpressionIO::ConfigurationType::Current)
+        ;
```
Comment (matekelemen) [inline kratos/python/add_expression_io_to_python.cpp]: I definitely remember an enum for `Configuration`s somewhere. I think I used it ages ago in some point or vertex output.

Anyway, it should rather be in a standard location; probably around [`DataLocation`](https://github.com/KratosMultiphysics/Kratos/pull/11254).

Comment (philbucher) [reply kratos/python/add_expression_io_to_python.cpp]: It exists already

I think you even added it :)

Comment (sunethwarna) [reply kratos/python/add_expression_io_to_python.cpp]: Yeah, it exists... and I even used the same name same exact everything as in the one which exists.... coincidence ? ;)

Comment (sunethwarna) [reply kratos/python/add_expression_io_to_python.cpp]: done :)

File: kratos/expression/expression_io_utils.h (hunk 4 above)
Comment (sunethwarna) [reply kratos/expression/expression_io_utils.h]: done :)

**Comment by sunethwarna** [comment]
@RiccardoRossi I am going to start writing the user/developer documentation this week :).

**Review by matekelemen — APPROVED** [review]

**Pull request merged**